sam/ancientDiploidBase.go

```diff
@@ -0,0 +1,201 @@
+	case AG:
+		firstTerm = pointFiveMinusEpsilonOverThreePlusLambdaOverTwoLikelihoodExpression(aCount, epsilon, lambda, cache)
+		secondTerm = epsilonOverThreeLikelihoodExpression(cCount+tCount, epsilon, cache)
+		thirdTerm = pointFiveMinusEpsilonOverThreeMinusLambdaOverTwoLikelihoodExpression(gCount, epsilon, lambda, cache)
+	case AT:
+		firstTerm = pointFiveMinusEpsilonOverThreeLikelihoodExpression(aCount+tCount, epsilon, cache)
+		secondTerm = epsilonOverThreeLikelihoodExpression(cCount+gCount, epsilon, cache)
+	case CC:
+		firstTerm = epsilonOverThreeLikelihoodExpression(aCount+gCount, epsilon, cache)
+		secondTerm = oneMinusEpsilonMinusLambdaLikelihoodExpression(cCount, epsilon, lambda, cache)
+		thirdTerm = epsilonOverThreePlusLambdaLikelihoodExpression(tCount, epsilon, lambda, cache)
+	case CG:
+		firstTerm = epsilonOverThreePlusLambdaOverTwoLikelihoodExpression(aCount, epsilon, lambda, cache)
+		secondTerm = pointFiveMinusEpsilonOverThreeMinusLambdaOverTwoLikelihoodExpression(cCount+gCount, epsilon, lambda, cache)
+		thirdTerm = epsilonOverThreePlusLambdaOverTwoLikelihoodExpression(tCount, epsilon, lambda, cache)
+	case CT:
+		firstTerm = epsilonOverThreeLikelihoodExpression(aCount+gCount, epsilon, cache)
+		secondTerm = pointFiveMinusEpsilonOverThreeMinusLambdaOverTwoLikelihoodExpression(cCount, epsilon, lambda, cache)
+		thirdTerm = pointFiveMinusEpsilonOverThreePlusLambdaOverTwoLikelihoodExpression(tCount, epsilon, lambda, cache)
+	case GG:
+		firstTerm = epsilonOverThreePlusLambdaLikelihoodExpression(aCount, epsilon, lambda, cache)
+		secondTerm = epsilonOverThreeLikelihoodExpression(cCount+tCount, epsilon, cache)
+		thirdTerm = oneMinusEpsilonMinusLambdaLikelihoodExpression(gCount, epsilon, lambda, cache)
+	case GT:
+		firstTerm = epsilonOverThreePlusLambdaOverTwoLikelihoodExpression(aCount, epsilon, lambda, cache)
+		secondTerm = epsilonOverThreeLikelihoodExpression(cCount, epsilon, cache)
+		thirdTerm = pointFiveMinusEpsilonOverThreeMinusLambdaOverTwoLikelihoodExpression(gCount, epsilon, lambda, cache)
+		fourthTerm = pointFiveMinusEpsilonOverThreeLikelihoodExpression(tCount, epsilon, cache)
+	case TT:
+		firstTerm = epsilonOverThreeLikelihoodExpression(aCount+cCount+gCount, epsilon, cache)
+		secondTerm = oneMinusEpsilonLikelihoodExpression(tCount, epsilon, cache)
+	default:
+		log.Fatalf("Error: Unknown genotype: %v.\n", geno)
+	}
+	return logspace.Multiply(firstTerm, logspace.Multiply(secondTerm, logspace.Multiply(thirdTerm, fourthTerm)))
+}
+
+// epsilonOverThreeLikelihoodExpression is a helper function of ancientBaseLikelihood that calculates the expression
+// (epsilon/3.0)^count.
+// Checks the cache to see if this expression has already been calculated. Otherwise, updates the cache.
+func epsilonOverThreeLikelihoodExpression(count int, epsilon float64, cache AncientLikelihoodCache) float64 {
+	if count < len(cache.EpsilonOverThree) { // if the coverage is within the cache bounds
+		if cache.EpsilonOverThree[count] != 0 {
```

Comment:
Not a problem, but this logic is repeated in a few functions and could be shortened to:
if cache.Something[count] == 0 { // if cache miss 
     cache.Something[count] = math // put it in the cache
}
return cache.Something[count]